test/h-group/level-11.js

```diff
@@ -913,6 +913,96 @@ describe('guide principle', () => {
+		takeTurn(game, 'Donald plays b1', 'r4');
+
+		// Alice can't know whether the g1 > g2 finesse is real until she knows whether Donald was bluffed.
+		// She should play the 2 first to find out.
+		const action = take_action(game);
+		ExAsserts.objHasProperties(action, {type: ACTION.PLAY, target: game.state.hands[PLAYER.ALICE][2].order});
+		takeTurn(game, 'Alice plays r2 (slot 3)');
+
+		// Recognizing that Donald played b1 for the bluff, Alice is not finessed.
+		assert.equal(game.common.thoughts[game.state.hands[PLAYER.ALICE][0].order].finessed, false);
+		ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.BOB][3].order], ['g2']);
+	});
+
+	it(`understands an ambiguous bluff delaying a finesse`, () => {
+		const game = setup(HGroup, [
+			['xx', 'xx', 'xx', 'xx'],
+			['b3', 'y3', 'y2', 'g2'],
+			['r3', 'b3', 'p5', 'p4'],
+			['b1', 'r4', 'b2', 'y4']
+		], {
+			level: { min: 11 },
+			play_stacks: [1, 0, 0, 0, 0],
+			starting: PLAYER.BOB
+		});
+		takeTurn(game, 'Bob clues 2 to Alice (slot 3)');
+		ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.ALICE][2].order], ['b2', 'r2']);
+
+		// Cathy could be bluffing out b1 or finessing a g1 in our hand.
+		takeTurn(game, 'Cathy clues green to Bob');
+
+		// Donald could be playing into the b1 > b2 finesse or a g2 bluff from Cathy.
+		takeTurn(game, 'Donald plays b1', 'r4');
+
+		// Alice can't know whether the g1 > g2 finesse is real until she knows whether Donald was bluffed.
+		// She should play the 2 first to find out.
+		const action = take_action(game);
+		ExAsserts.objHasProperties(action, {type: ACTION.PLAY, target: game.state.hands[PLAYER.ALICE][2].order});
+		takeTurn(game, 'Alice plays b2 (slot 3)');
+
+		// Recognizing that Donald played b1 for the finesse, Alice is finessed.
+		ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.ALICE][1].order], ['g1']);
+		assert.equal(game.common.thoughts[game.state.hands[PLAYER.ALICE][1].order].finessed, true);
+	});
+
+	it(`understands delaying on an ambiguous bluff`, () => {
+		const game = setup(HGroup, [
+			['g1', 'r5', 'b2', 'g3'],
+			['xx', 'xx', 'xx', 'xx'],
+			['r3', 'b3', 'p5', 'p4'],
+			['b1', 'r4', 'b2', 'y4']
+		], {
+			level: { min: 11 },
+			play_stacks: [1, 0, 0, 0, 0],
+			starting: PLAYER.BOB
+		});
+		game.state.ourPlayerIndex = 1;
+		takeTurn(game, 'Bob clues 2 to Alice');
+		ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.ALICE][2].order], ['b2', 'r2']);
+
+		// Cathy finessing out g1 in our hand.
+		takeTurn(game, 'Cathy clues green to Bob (slot 4)');
+		assert.equal(game.common.thoughts[game.state.hands[PLAYER.ALICE][0].order].finessed, true);
+		ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.BOB][3].order], ['g1', 'g2']);
+
+		takeTurn(game, 'Donald plays b1', 'r4');
+		takeTurn(game, 'Alice plays b2', 'b4');
+
+		// Alice is still finessed.
```

Comment:
This part is still broken, Bob doesn't currently recognize that Alice had any reason to wait.

Reply:
This seems to be because `find_connecting` searches backwards from the target finding Alice so the potential "bluff" is never considered. Perhaps we'll need to add an additional check to indicate that Alice might wait. Perhaps for these cases instead of separately finding the finesse connection we could use the already known ambiguous finesse in Donald's hand.

Reply:
I don't think I fully understand your proposed solutions, but I'm not actually sure the current modelling system is capable of handling this. Despite having independent "thoughts" for each player, it actually only uses the common interpretation + extra elim notes from seeing other people's cards, so there is no support for doing things like one player having a completely different interpretation from others. The common interpretation also is not really a good base; it's very difficult to figure out what is collectively known around finesses, so I just used the correct interpretation + any symmetric interpretations the target could have. Alice is unable to figure out the correct interpretation and she isn't the target, so I think it'll be hard to correctly model her with the current setup.

`find_connecting()` searches backwards from the target because anyone searching for a missing identity will always wait for the player closest to the target (as an *Ambiguous Finesse*). Thus, it doesn't make sense that Donald will play b1 as a bluff into g2, because Donald can see Alice's g1 on finesse position. Of course, this argument falls apart when you need to consider Alice's perspective, but that's exactly what the current common interpretation doesn't account for.

If one of your solutions works as a band-aid, that'd be great! But I think it might be shaky, with the bot failing to respond correctly to slightly different situations.

Reply:
In f08358e I realized that the correct order is actually going backwards from the giver, not the target... Also, there was a similar situation where a bluff was successfully interpreted despite a possible finesse connection existing (see "attempts to prompt when considering whether a bluff is valid"), so maybe this is possible.